### ISSUE TITLE
Generate new credentials on every bind

### ISSUE THREAD

**Issue by mogul** [18F/aws-broker#12]
Cycling the credentials an app uses to access its database in case of compromise should be as easy as unbinding and rebinding the service and the app. Currently `aws-broker` doesn't re-generate credentials when you do this, and it should.

Alternatively, we can evaluate switching to https://github.com/cf-platform-eng/rds-broker which does have this behavior.

**Comment by sarneaud** [comment]
I'm working at a site that's been using https://github.com/cf-platform-eng/rds-broker and is considering this broker.

I've come across a couple of problems with how that broker works right now:

* There's a risk of losing access to your old crediantials (and data) if you rebind for any reason.
* It's really hard to coordinate access among multiple apps.  (GRANT is almost useless, for example.)

These are really problems with usernames being unstable (they get randomly generated on each bind).

I think an approach that would work would be to randomly generate passwords on bind, but using a per-app username from the bind parameters (with a reasonable default).

**Comment by sarneaud** [comment]
I've been reviewing the code and figuring what changes are needed to make it work.  Before I make a PR, is this something that's actually wanted by the maintainers?  Here's the behaviour I'm talking about:

* A database is created on instance creation, and deleted on instance deletion.
* A database user is created on app bind, and deleted on unbind.
* The username can be specified as a parameter to the binding.

**Comment by mogul** [comment]
Sorry for the delayed response, @sarneaud... Yes, this is something we want, and we'd appreciate any pull-requests!

**Comment by mogul** [comment]
Just a note for anyone else happening across this: The implication of the current behavior is that every app in a particular space with a service instance will get the same credentials.

**Comment by mogul** [comment]
From https://github.com/18F/cg-s3-service-broker

> A service provisioning call will create an S3 bucket, an IAM group, and an IAM Policy to provide access controls on the bucket. A binding call will create an IAM user, generate access keys, and add it to the bucket's group. Unbinding and deprovisioning calls will delete all resources created.

That's how this broker should work as well.

**Comment by mogul** [comment]
@sarneaud, did you ever end up working on this? We got bitten by this (the difficulty of cycling the credential after someone accidentally committed it to a public repository) again today. We don't have a lot of room among our other prioritized stuff to work on it, but if I know you're not working on it, maybe I'll use that to push it up the list.

**Comment by jmcarp** [comment]
@sarneaud: building usernames based on app names sounds reasonable, but what's the motivation for overriding usernames in bind parameters? Also, I think it would be less work to modify the bind behavior of https://github.com/cf-platform-eng/rds-broker than to add it here.

@mogul: if/when we allocate time to this again, I'm leaning toward using https://github.com/cf-platform-eng/rds-broker instead of continuing on this codebase. The pivotal project already has a version of the behavior this issue is requesting (which can easily be modified based on @sarneaud's proposal). It also uses brokerapi rather than martini / gin, which means less boilerplate; it uses asynchronous operations on provision instead of erroring while pending (which is what this broker does); it supports more engines, including aurora clusters; it respects some useful optional arguments on provision and update; and it has decent test coverage. If we decide to stick with this version, that list goes into our backlog. @dlapiduz and @jcscottiii: am I being too hard on our code?

**Comment by mogul** [comment]
That all sounds desirable, and the likelihood that wed sink our own resources into bringing all of that to this codebase is low. Are there any advantages for this one?

**Comment by jmcarp** [comment]
Unfortunately, yes: our broker has a shared service, and the alternative only has dedicated services. But given that our kubernetes broker includes postgres and mysql, maybe we shift the free plan there.

**Comment by jcscottiii** [comment]
Yeah, the broker is outdated now. Back then, when we started there were no libraries and the CF api was basic. But now, there are helper libraries (e.g. https://github.com/pivotal-cf/brokerapi) which keeps up to date with the CF api now.

Unless, we are still trying to continue to make the aws broker (the broker of all things AWS), I would say we could sunset support for this.

Someone needs to reach out to the cf-platform-eng maintainers to see who is actively maintaining it so we don't have to keep a fork. I only ask since they haven't touched it since last nov.

**Comment by jmcarp** [comment]
Relevant discussion in https://github.com/cf-platform-eng/rds-broker/issues/2, but unfortunately, it looks like that repo also isn't going to get any maintenance--I think we're stuck doing our own maintenance whichever broker we use. It might be worth asking pivotal to transfer the brokers they're not maintaining to cloudfoundry-community.

**Comment by mogul** [comment]
Good idea, and I've prompted for that in the issue.

On the topic of a shared plan being provided via the kubernetes broker... Is it actually possible to offer plans within a single service that refer to two separate brokers?! If so I'm surprised.

**Comment by jmcarp** [comment]
Thanks @mogul! And yeah, what I meant was that we could offer dedicated databases through the aws/rds-broker and shared through the kubernetes broker. So the plans would appear under different services--which could be a little confusing, but IMO is better than reimplementing the shared rds service in https://github.com/cf-platform-eng/rds-broker.

**Comment by sarneaud** [comment]
Sorry about the late response from me.

No, I'm not working on this now, but I'll chip in my 2c.

Username==appname is enough for most deployments, but being able to override the username is useful for complex deployments where something is split into multiple CF apps but is still one thing from an access control point of view.  Here are two concrete examples I know of: applications run as two apps for high availability (CF doesn't support rolling restarts and shuts down apps for upgrade) and (I've seen this on another platform) applications with a two-stage production rollout for extra safety -- first to an app that runs a few instances, then globally if no problems are detected.  The alternative is having a username per app and managing credentials with GRANT or equivalent.

Unfortunately, when I looked more closely at the CF API, I didn't see any way to get the app name on bind.  Hopefully I just didn't look hard enough.  Also, although the API supports custom bind parameters, the command line tool currently only supports them as command line parameters, not in the manifest.yml file, which makes things a bit impractical.

Sorry I can't help any more.  Good luck whatever path you decide to go down.

**Comment by mogul** [comment]
> > It might be worth asking pivotal to transfer the brokers they're not maintaining to cloudfoundry-community.
> Good idea, and I've prompted for that in the issue.

Good news, Pivotal moved it over. 👍
